File: star-zmq/python/star_zmq.py
"""Local-only transport binding. Actor semantics and JSON validation live upstream."""
from __future__ import annotations

import re
import threading

import zmq

MAX_PAYLOAD = 1_048_576
_LOCAL_TCP = re.compile(r"tcp://(?:127\.0\.0\.1|\[::1\]):([0-9]{1,5})\Z", re.ASCII)


class TransportLimitError(ValueError):
    """An endpoint, frame count or payload bound was violated."""


class SocketOwnerError(RuntimeError):
    """A socket was used from another thread."""


def local_endpoint(endpoint: str) -> bool:
    if not isinstance(endpoint, str) or not 1 <= len(endpoint) <= 1024 or "\0" in endpoint:
        return False
    for prefix in ("ipc://", "inproc://"):
        if endpoint.startswith(prefix) and len(endpoint) > len(prefix):
            return True
    match = _LOCAL_TCP.fullmatch(endpoint)
    return match is not None and 1 <= int(match[1]) <= 65535


class Dealer:
    """One explicitly addressed peer. The caller owns the shared process context.

    A successful send only means queued to libzmq. This class does not retry,
    authorize identities, implement actors, serialize objects or promise delivery.
    """

    def __init__(
        self, context: zmq.Context, endpoint: str, identity: bytes, *,
        timeout_ms: int = 1000, payload_limit: int = MAX_PAYLOAD,
    ) -> None:
        if not local_endpoint(endpoint):
            raise TransportLimitError("Only IPC/inproc/literal loopback endpoints are allowed")
        if not isinstance(identity, bytes) or not 1 <= len(identity) <= 255 or identity[0] == 0:
            raise TransportLimitError("Identity must contain 1..255 bytes, without an initial NUL")
        if type(timeout_ms) is not int or not 1 <= timeout_ms <= 60_000:
            raise TransportLimitError("Timeout must be 1..60000 milliseconds")
        if type(payload_limit) is not int or not 1 <= payload_limit <= 16_777_216:
            raise TransportLimitError("Invalid payload limit")
        self._owner = threading.current_thread()
        self._limit = payload_limit
        self._socket: zmq.Socket | None = context.socket(zmq.DEALER)
        try:
            self._socket.setsockopt(zmq.LINGER, 0)
            self._socket.setsockopt(zmq.SNDHWM, 64)
            self._socket.setsockopt(zmq.RCVHWM, 64)
            self._socket.setsockopt(zmq.RCVTIMEO, timeout_ms)
            self._socket.setsockopt(zmq.SNDTIMEO, timeout_ms)
            self._socket.setsockopt(zmq.MAXMSGSIZE, max(255, payload_limit))
            self._socket.setsockopt(zmq.IMMEDIATE, 1)
            self._socket.setsockopt(zmq.ROUTING_ID, identity)
            self._socket.connect(endpoint)
        except BaseException:
            self.close()
            raise

    def _owned(self) -> zmq.Socket:
        if threading.current_thread() is not self._owner:
            raise SocketOwnerError("Socket belongs to its creating thread")
        if self._socket is None:
            raise RuntimeError("Socket is closed")
        return self._socket

    def send(self, payload: bytes) -> None:
        socket = self._owned()
        if not isinstance(payload, bytes) or len(payload) > self._limit:
            raise TransportLimitError("Expected bounded immutable bytes")
        socket.send(payload, copy=True)

    def receive(self) -> bytes:
        socket = self._owned()
        # recv_into reports the full native frame length without allocating that length.
        buffer = bytearray(self._limit)
        size = socket.recv_into(buffer)
        try:
            if size > self._limit or socket.getsockopt(zmq.RCVMORE):
                raise TransportLimitError("Oversized or unexpected multipart message")
            return bytes(buffer[:size])
        except BaseException:
            self.close()
            raise

    def close(self) -> None:
        if threading.current_thread() is not self._owner:
            raise SocketOwnerError("Close the socket on its owner thread")
        if self._socket is not None:
            self._socket.close(linger=0)
            self._socket = None

    def __enter__(self) -> Dealer:
        self._owned()
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

File: star-zmq/nim/star_zmq.nim
## A single-threaded external peer binding, not an alternate StarLang runtime.
## Deliberately local-only until the authenticated federation port is available.
import std/strutils

when compileOption("threads"):
  {.error: "This first peer binding requires --threads:off; never share ZMQ sockets.".}

const
  DefaultLibrary = when defined(windows): "libzmq.dll"
                   elif defined(macosx): "libzmq.dylib"
                   else: "libzmq.so.5"
  StarZmqLibrary {.strdefine.} = DefaultLibrary
  PayloadLimit* = 1_048_576

proc ctxNew(): pointer {.cdecl, importc: "zmq_ctx_new", dynlib: StarZmqLibrary.}
proc ctxTerm(ctx: pointer): cint {.cdecl, importc: "zmq_ctx_term", dynlib: StarZmqLibrary.}
proc socketNew(ctx: pointer; kind: cint): pointer {.cdecl, importc: "zmq_socket", dynlib: StarZmqLibrary.}
proc socketClose(socket: pointer): cint {.cdecl, importc: "zmq_close", dynlib: StarZmqLibrary.}
proc socketConnect(socket: pointer; endpoint: cstring): cint {.cdecl, importc: "zmq_connect", dynlib: StarZmqLibrary.}
proc setOpt(socket: pointer; option: cint; data: pointer; size: csize_t): cint {.cdecl, importc: "zmq_setsockopt", dynlib: StarZmqLibrary.}
proc getOpt(socket: pointer; option: cint; data: pointer; size: ptr csize_t): cint {.cdecl, importc: "zmq_getsockopt", dynlib: StarZmqLibrary.}
proc sendRaw(socket: pointer; data: pointer; size: csize_t; flags: cint): cint {.cdecl, importc: "zmq_send", dynlib: StarZmqLibrary.}
proc recvRaw(socket: pointer; data: pointer; size: csize_t; flags: cint): cint {.cdecl, importc: "zmq_recv", dynlib: StarZmqLibrary.}
proc errorNumber(): cint {.cdecl, importc: "zmq_errno", dynlib: StarZmqLibrary.}
proc errorString(code: cint): cstring {.cdecl, importc: "zmq_strerror", dynlib: StarZmqLibrary.}

type
  ZmqError* = object of CatchableError
    code*: int
  Dealer* = ref object
    context: pointer
    socket: pointer

var activeDealer: bool

proc checked(value: cint; operation: string): cint =
  if value < 0:
    let code = errorNumber()
    var failure = newException(ZmqError, operation & ": " & $errorString(code))
    failure.code = int(code)
    raise failure
  value

proc localEndpoint*(endpoint: string): bool =
  if endpoint.len == 0 or endpoint.len > 1024 or '\0' in endpoint:
    return false
  if endpoint.startsWith("ipc://") and endpoint.len > 6: return true
  # inproc is process-local and cannot connect this child to a Lisp parent.
  for prefix in ["tcp://127.0.0.1:", "tcp://[::1]:"]:
    if endpoint.startsWith(prefix):
      let port = endpoint[prefix.len .. ^1]
      if port.len < 1 or port.len > 5: return false
      for c in port:
        if c notin {'0'..'9'}: return false
      let number = parseInt(port)
      return number >= 1 and number <= 65535
  false

proc close*(dealer: Dealer) =
  if dealer.isNil: return
  if dealer.socket != nil:
    discard checked(socketClose(dealer.socket), "close")
    dealer.socket = nil
  if dealer.context != nil:
    discard checked(ctxTerm(dealer.context), "context-term")
    dealer.context = nil
    activeDealer = false

proc openDealer*(endpoint, identity: string; timeoutMs = 5000; lingerMs = 0): Dealer =
  if not localEndpoint(endpoint) or identity.len notin 1..255 or
      identity[0] == '\0' or timeoutMs notin 1..60000 or lingerMs notin 0..5000:
    raise newException(ValueError, "Invalid local endpoint, identity, timeout, or linger")
  if activeDealer:
    raise newException(ValueError, "One peer/context per process in this binding")
  new(result)
  result.context = ctxNew()
  if result.context == nil:
    discard checked(-1, "context-new")
  activeDealer = true
  try:
    result.socket = socketNew(result.context, 5) # DEALER
    if result.socket == nil: discard checked(-1, "socket")
    for entry in [(17.cint, cint(lingerMs)), (23.cint, 64.cint), (24.cint, 64.cint),
                  (27.cint, cint(timeoutMs)), (28.cint, cint(timeoutMs)), (39.cint, 1.cint)]:
      var value = entry[1]
      discard checked(setOpt(result.socket, entry[0], addr value, csize_t(sizeof(value))), "setsockopt")
    var maxBytes = int64(PayloadLimit)
    discard checked(setOpt(result.socket, 22, addr maxBytes, csize_t(sizeof(maxBytes))), "max-message-size")
    discard checked(setOpt(result.socket, 5, cast[pointer](identity.cstring), csize_t(identity.len)), "routing-id")
    discard checked(socketConnect(result.socket, endpoint.cstring), "connect")
  except:
    result.close()
    raise

proc requireOpen(dealer: Dealer) =
  if dealer.isNil or dealer.socket == nil:
    raise newException(ValueError, "Dealer is closed")

proc send*(dealer: Dealer; payload: string) =
  dealer.requireOpen()
  if payload.len > PayloadLimit:
    raise newException(ValueError, "Payload limit exceeded")
  let written = checked(sendRaw(dealer.socket, cast[pointer](payload.cstring), csize_t(payload.len), 0), "send")
  if written != payload.len:
    dealer.close()
    raise newException(ValueError, "Unexpected short send")

proc receive*(dealer: Dealer): string =
  dealer.requireOpen()
  result = newString(PayloadLimit)
  # No read past the allocated buffer: the returned native length can be larger.
  let size = checked(recvRaw(dealer.socket, addr result[0], csize_t(result.len), 0), "receive")
  try:
    if size > PayloadLimit:
      raise newException(ValueError, "Truncated/oversized payload")
    var more: cint
    var optionSize = csize_t(sizeof(more))
    discard checked(getOpt(dealer.socket, 13, addr more, addr optionSize), "receive-more")
    if more != 0:
      raise newException(ValueError, "Unexpected multipart payload")
    result.setLen(int(size))
  except:
    dealer.close()
    raise
